doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,291 @@
+- A service is a series of one or more API calls that must all succeed
+- A extra API call is needed after the service succeeds, that should return 1 if the service is approved.
+
+#### Solution 1 (Using an indicator everywhere)
+
+Use a per thread global counter that is incremented when an algorithm is approved.  AWS/google have added this in places where a service is at the point of completing its output (e.g. digest_final). This design is complicated by the fact that a service may call another service (e.g HMAC using SHA256) that also wants to increment the approved counter. To get around this issue they have a second variable that is used for nesting. If the variable is non zero then the approved counter doesnt increment. This also allows non security relevant functions to not increment the approved count. Another variation of this would be to use flags instead of a counter.
+
+##### Cons
+- At the fips provider level this would require some plumbing to go from the core to the fips provider, which seems overly complex.
+- The query can only be done after the output is set.
+- The indicator code would end up having to be set in different places depending on the algorithm after the output
+is finalized. This would be fairly messy as the point where it is called is set could be different for different algorithms.
+- The locking increment seems messy.
+
+
+#### Proposed Solution (Using an indicator everywhere)
+
+Add a OSSL_PARAM getter to each provider algorithm.
+By default if the getter is not handled then it would return not approved.
+
+##### Pros
+- The code is easier to find since it is part of the get_ctx_params function.
+- The getter can be called at any point after all the setting is done.
+
+Any fips algorithm that is approved would then need a setter that at a minimum contains code similar to the following
+
+~~~
+int ossl_xxx_fips_approved(void)
+{
+#ifdef FIPS_MODULE
+    return 1; /* conditional code would go here for each algorithm if required *
+#else
+    return 0;
+#endif
+}
+~~~
+
+and in the algorithms get_ctx() function
+
+~~~
+int xxx_get_fips_approved(OSSL_PARAM params[])
+{
+    p = OSSL_PARAM_locate(params, OSSL_FIPS_PARAM_APPROVED);
+    if (p != NULL && !OSSL_PARAM_set_int(p, ossl_xxx_fips_approved()))
+        return 0;
+	return 1;
+}
+~~~
+
+#### API’s that would be used to support this are
+
+- EVP_PKEY Keygen, Encryption, Signatures, Key Exchange, KEM
+~~~ 
+EVP_PKEY_CTX_get_params(ctx, );
+~~~
+(Note that this would mean you couldnt use functions that hide the ctx such as EVP_PKEY_Q_keygen()!)
+
+- Ciphers
+~~~
+EVP_CIPHER_CTX_get_params()
+~~~
+
+- Digests
+~~~
+EVP_MD_CTX_get_params()
+~~~
+
+- KDF’s
+~~~
+EVP_KDF_CTX_get_params()
+~~~
+
+- MAC’s
+~~~
+EVP_MAC_CTX_get_params()
+~~~
+
+- RAND
+~~~
+EVP_RAND_CTX_get_params()
+~~~
+
+#### Backwards Compatability..
+
+Previous providers do not support this operation, so they will return not approved if they are not handled.
```

Comment:
Maybe????

Reply:
The thinking being that it wont handle the OSSL_PARAM so it is not approved.

Reply:
I believe the FIPS provider has many more algorithms which are approved rather than non-approved. Would it not be easier to have an OSSL_PARAM indicating the *non-approved* status? If this parameter is present and set to `true`, the service would be non-approved. The absence of the parameter, or (if it is present) a value of `false` would mean the service is approved. That way you don't have to add the parameter to every single API, but only the ones where non-approvedness is possible.

On the other hand, this becomes more difficult to explain in the Security Policy.

Reply:
I was thinking this applies to all providers, and in that case undefined is better to mean not approved, especially if we apply this to any provider that doesnt understand this. The set of algorithms of approved algorithms will always  be a subset of all algorithms.

Reply:
From that point of view that's correct – however keep in mind that only the FIPS provider is certified anyway. So from a certification point of view, it doesn't matter what values other providers do or don't return for this param. If you want to be FIPS compliant, you already have to make sure that you are using the FIPS provider.

Red Hat went for the "if it doesn't have an indicator and is in the FIPS provider, it's approved" default.